newline edge case fixes

bartbutler opened the pull request:
A couple minor newline fixes for cases with nested divs and fixing the trailing br tag stripping if the tag is followed by some whitespace.

soundasleep commented on src/Html2Text.php:
```diff
@@ -295,19 +316,35 @@ static function iterateOverNode($node, $prevName = null, $in_pre = false, $is_of
+			}
+
+			// Remove trailing whitespace, important for the br check below
+			while($trailing_whitespace-- !== 0) {
```
I would rather have `while ($trailing_whitespace-- > 0)` to prevent any possible infinite loop.

soundasleep commented on tests/newlines.html:
```diff
@@ -3,11 +3,7 @@
 <div>
 Hello
 <br>
-</div>
```
I'm concerned that you had to change existing behaviour for this change, as this package is used by many thousands of people, I would like the existing behaviour to continue (as long as it's reasonable). Can you revert the changes to the existing test cases and instead add new test cases illustrating the new behaviour?

Alternatively, can you provide screenshots showing that the original test was incorrect c.f. rendering HTML in a browser?

soundasleep commented on src/Html2Text.php:
```diff
@@ -263,10 +274,16 @@ static function iterateOverNode($node, $prevName = null, $in_pre = false, $is_of
 				// add one line
 				$output = "\n";
 				break;
+			case "div":
```
Please put in an empty line between the break and the next case statement, so it's consistent with the rest of the file

soundasleep commented on tests/nbsp.txt:
```diff
@@ -1 +1 @@
-hello world & people < > &NBSP;
\ No newline at end of file
+hello   world & people < > &NBSP;
```
I think you've regressed some behaviour here

bartbutler replied:
No, actually I've corrected it.
<img width="262" alt="screen shot 2017-08-09 at 1 24 28 am" src="https://user-images.githubusercontent.com/5491415/29112095-930c9be0-7ca1-11e7-979a-32b43d20bb11.png">


Rendered HTML screenshot (Firefox):

soundasleep replied:
Ahh nice catch!

soundasleep commented:
Thank you for the PR - I've left some comments in the review that I'd like addressed first, particularly around breaking existing behaviour.

bartbutler commented:
I have addressed your comments. I've split the nested div stuff into a separate nested div test and restored the old newline test, which continues to pass with no changes. The newline behavior with divs/nested divs was just wrong in certain cases--all the div and br code changes are basically bugfixes (for instance, the br trimming at the end of a div or other container tag was dependent on whether whitespace was present between the br tag and the closing tag, which is not at all what we want).

These changes did make two minor newline changes to the full-email test. Converting that email to plain text requires some imagination but it is safe to say the new text version is not less correct than the old one was, and I'm inclined to defer to the code which renders other nested divs correctly.

The non-breaking space treatment of the old code was just wrong, as was the test (see comment). Lots of documents use non-breaking spaces as a way to indent HTML text, and the old code would just convert them to spaces and then strip them like any other HTML whitespace. The new code protects leading and internal non-breaking spaces to preserve indentation and other desired formatting which still removing (useless) trailing whitespace of all types and converting non-breaking spaces to ASCII spaces post-render to avoid unnecessary Unicode issues. This really shouldn't be an issue for any existing users--the rendering is more correct and the impact is pretty minimal--after all, the office test case has a bunch of non-breaking spaces and that test continues to pass with no changes at all.

soundasleep commented:
This is great! Thank you so much for the PR.

ahfeel commented:
Hello !

I think this pull request introduces a major issue with removal of <br> nodes when they are the last child of a node that also contained text. Here's a very simple example of a big regression (this is stuff I'm receiving from bad html emails) :

```
<font size="+1">Vikings: Wolves of Midgard<br></font><font size="+1">Valkyria Chronicles<br>
<br>
World Of Warcraft Battlechest</font>
```
The expected output would be

```
Vikings: Wolves of Midgard
Valkyria Chronicles

World Of Warcraft Battlechest
```

The actual output with the changes introduced in this commit are:

```
Vikings: Wolves of MidgardValkyria Chronicles

World Of Warcraft Battlechest
```

@bartbutler Could you please explain in which exact case this is necessary for you ? It may generate additional linebreaks but in any case all the unnecessary linebreaks are removed at the end of the processing.

ahfeel commented:
Huh, actually after reviewing the original code, the behaviour was already broken. I'll just open a new issue :)
